### PR TITLE
Add syntax trees

### PR DESCRIPTION
**Pull request by paltman**
Screen Recording 2020-09-09 at 03 37 38 PM

### CODE REVIEW

File: packages/vue-treant/src/Treant.vue [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        const nodes = this.$el.querySelectorAll('.node');
+        nodes.forEach(node => {
+          node.addEventListener('mouseenter', this.onNodeEnter);
+          node.addEventListener('mouseleave', this.onNodeLeave);
+        });
+      },
+      setWidth() {
+        this.width = `${this.$parent.$el.clientWidth}px`;
+      },
+      onResize() {
```
Comment (jacobwegner) [inline packages/vue-treant/src/Treant.vue:39]: What about attaching a watcher to sidebar toggles?

e.g, if the user toggles a sidebar on / off, force redraw?

File: packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        const tree = {id: 0, value: null, headId: null, relation: null, children: []};
+
+        words.forEach(word => {
+          if (word.headId === 0) {
+            tree.children.push(word);
+          } else {
+            wordBank[word.headId].children.push(word);
+          }
+        });
+        return {
```
Comment (jacobwegner) [inline packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:103]: two comments here:

1) If you paginate to the next passage, you get an empty "Root":

image


I think that we need to check  in`<EmptyMessage v-else-if="!data" />` off of something other than data that has a falsey value..likely checking the number of nodes on the edge, for example

2) related to the above, this structure assumes that there is only _one_ syntax tree for a given passage.  I will expand the dataset and ping back here when deployed so we can have a data stub with multiple trees in a single passage.

File: packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        `;
+      },
+    },
+  };
+</script>
+
+<style lang="scss" scoped>
+  .reader {
+    width: 100%;
+    // max-width: 600px;
```
Comment (jacobwegner) [inline packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:131]: Kill the comment?
```suggestion
```

File: packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue (hunk 2 above)
Comment (jacobwegner) [reply packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:103]: @paltman I've loaded more data into ATLAS.

If you paginate from 1.7 to 1.8, there seems to be a reactivity issue:

https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.7

The sentence / words changes, but the graph does not:

https://share.getcloudapp.com/04uqZb1W

Comment (jacobwegner) [reply packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:103]: `1.9` should have two trees:

https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.9

https://www.perseids.org/tools/arethusa/app/#/perseids?chunk=3&doc=28550

https://www.perseids.org/tools/arethusa/app/#/perseids?chunk=4&doc=28550

File: packages/vue-treant/src/Treant.vue (hunk 1 above)
Comment (paltman) [reply packages/vue-treant/src/Treant.vue:39]: Yeah, I think we need to do that in addition to what I'm doing here.

File: packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue (hunk 2 above)
Comment (paltman) [reply packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:127]: How should we show multiple trees?

**Comment by paltman** [comment]
Added some more this morning:

Screen Recording 2020-09-11 at 08 55 18 AM

File: packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue (hunk 2 above)
Comment (jacobwegner) [reply packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:127]: @paltman If you look at the arethusa links above, they are paginating there by "chunk" (that is, sentences).

When we groomed the syntax trees (and developed the ATLAS query), we're returning sentence(s) that map to the current passage, so there could be more than one.

This is similar to what we're doing with the metrical display mode or the "old" sentence alignments:

_1.1-1.5: 1 alignment_

image
<br />

_1.6-1.10: 4 alignments_

image
<br />

_1.11-1.15: 2 alignments_

image
<br />


(as an aside, folio image mode _should_ work like the sentence alignments, but is rendering just the first folio image; I'll capture an issue for further elaboration on that.)

I can think of a couple of way to handle multiple sentences for a single passage:

1) Short term, render the first tree, and provide an indication in the UX that additional _n_ trees are not being shown.  The `words` and `tree` data structures you're building would need to _only_ parse data from the first node returned if there are multiple trees.

2) Ideally, we're rendering multiple sentences within the reader mode.  You build up a list of `words` and `tree` data structures, and you loop through "rows" (just like the metrical mode).

```
# # # # # # # # # # # # # # # # # # # # # # # #
# sentence 1
# graph 1
# # # # # # # # # # # # # # # # # # # # # # # #

# # # # # # # # # # # # # # # # # # # # # # # #
# sentence2
# graph2
# # # # # # # # # # # # # # # # # # # # # # # #
```

If you toggled only sentences, you'd only see sentences, only graphs, only the graphs etc.

I'd suggested on our grooming call that we might want an accordion like toggle on the either the "row" or maybe just the "graph" so you could quickly collapse / expand the graph (e.g., the container, not the "nodes" in the graph):

_Figure A: expand/collapse entire rows_

```
# # # # # # # # # # # # # # # # # # # # # # # #
# [expanded \/ ]
# sentence1
# graph1
# # # # # # # # # # # # # # # # # # # # # # # #

# # # # # # # # # # # # # # # # # # # # # # # #
# [collapsed  > ]
# sentence1
# graph1
# # # # # # # # # # # # # # # # # # # # # # # #
```

_Figure B: expand/collapse only graphs_

```
# # # # # # # # # # # # # # # # # # # # # # # #
# sentence1
# graph1 [expanded \/ ]
# # # # # # # # # # # # # # # # # # # # # # # #

# # # # # # # # # # # # # # # # # # # # # # # #
# sentence2
# graph2 [collapsed > ]
# # # # # # # # # # # # # # # # # # # # # # # #
```

I think _Figure B_ would be better UX

Comment (jtauber) [reply packages/reader-syntax-trees-mode/src/SyntaxTreesModeReader.vue:127]: I agree, Figure B would work better. Then you can keep the context of the text while viewing just a single tree.

**Comment by jacobwegner** [comment]
@jtauber you mentioned in Slack you might want to play around in the browser; you can do so via this preview URL:

https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.1-1.7

**Comment by jacobwegner** [comment]
@paltman found one more small bug:

- Browse to https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.24-1.25
- Toggle on syntax tree display mode
- Paginate to 1.26-1.27 (https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.26-1.27)
- Shows Expand / Collapse toggle:
image


We likely need to tune the empty condition and populate a message using `EmptyMessage`

**Comment by paltman** [comment]
@jacobwegner done

**Comment by jacobwegner** [comment]
@paltman I found another bug where the the value of `expandAll` doesn't seem to persist correctly when navigating from the current passage to another passage.

A) Navigating to a next / previous passage

- Browse to https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.8-1.14

- Switch to "Syntax Tree" display mode

- Click "Expand All" to open all trees

- Navigate to 1.15-1.21

  - "Collapse All" is shown, but no trees are expanded
  - Clicking "Collapse All" doesn't seem to impact anything
  - Clicking "Expand All" expands all trees

https://share.getcloudapp.com/kpubXA7D


B) Navigating to another passage in another version

- Navigate to https://deploy-preview-21--scaife-viewer-frontend.netlify.app/explore-homer?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.8-1.14
- Enable syntax trees display mode
- Click to "Expand All" trees
- From the Library widget, navigate to Plato -> Apology - > Apology (Burnet)
- Expected: The first tree is expanded, and the `expandAll` label value is "Collapse All"
- Actual: The first tree is not expanded, and the label for `expandAll` is "Collapse All"

_Collapse all shown_    
image

_Clicking collapse all doesn't do anything_
image

_Clicking expand all finally opens all trees_
image

**Comment by paltman** [comment]
@jacobwegner fixed

**Review by jacobwegner — APPROVED** [review]
@paltman Bugfixes LGTM

**Pull request merged**